lua/rip-substitute/matches.lua

```diff
@@ -0,0 +1,221 @@
+				return nil, "[4]could not get replacement for " .. match.matchedText
+			end
+
+			local line = replaced[i]
+			local rowStr, colStr, replacementText = line:match("^(%d+):(%d+):(.*)")
+			---@type RipSubstituteMatch
+			match.replacementText = replacementText
+		end
+	end
+	return matches
+end
+
+---@param matches RipSubstituteMatch[]
+---@return RipSubstituteMatch | nil, string |nil
+function M.getClosestMatchAfterCursor(matches)
+	if #matches == 0 then return nil end
+	local state = require("rip-substitute.state").state
+	local cursor_row, cursor_col = unpack(vim.api.nvim_win_get_cursor(state.targetWin))
+	local closestMatch = nil -- Store the closest match found after cursor
+
+	cursor_row = cursor_row - 1
+	-- First, try to find a match after the cursor position
+	for _, match in ipairs(matches) do
+		local on_line_after = match.row > cursor_row
+		local on_same_line = match.row == cursor_row
+		local cursor_on_match = on_same_line and match.col <= cursor_col and match.col >= cursor_col
+		local on_same_line_after = not cursor_on_match and on_same_line and match.col > cursor_col
+
+		if on_line_after or cursor_on_match or on_same_line_after then
+			closestMatch = match
+			break -- Stop the loop if a match is found
+		end
+	end
+
+	-- If no match is found after cursor, search from the beginning of the file to the cursor
+	if not closestMatch then
+		for _, match in ipairs(matches) do
+			local on_line_before = match.row < cursor_row
+			local on_same_line_before = match.row == cursor_row and match.col < cursor_col
+
+			if on_line_before or on_same_line_before then
+				closestMatch = match
+				-- No break here; keep updating closestMatch to the last match before cursor
+			end
+		end
+	end
+
+	return closestMatch
+end
+
+---@param replacing boolean
+---@param newSelectedMatchIndex number
+local function updateSelectedMatchHighlight(replacing, newSelectedMatchIndex)
+	local state = require("rip-substitute.state").state
+	vim.api.nvim_buf_clear_namespace(
+		state.targetBuf,
+		state.incPreviewNs,
+		state.selectedMatch.row,
+		state.selectedMatch.row + 1
+	)
+	if replacing then
+		rg.highlightReplacement(state.selectedMatch, false, state.targetBuf, state.incPreviewNs)
+	else
+		rg.highlightMatch(
+			state.selectedMatch,
+			false,
+			state.targetBuf,
+			state.incPreviewNs,
+			state.selectedMatch.col + #state.selectedMatch.matchedText
+		)
+	end
+	state.selectedMatch = state.matches[newSelectedMatchIndex]
+	--TODO: this will probably clear too many highlights
+	vim.api.nvim_buf_clear_namespace(
+		state.targetBuf,
+		state.incPreviewNs,
+		state.selectedMatch.row,
+		state.selectedMatch.row + 1
+	)
+	if replacing then
+		rg.highlightReplacement(state.selectedMatch, false, state.targetBuf, state.incPreviewNs)
+	else
+		rg.highlightMatch(
+			state.selectedMatch,
+			true,
+			state.targetBuf,
+			state.incPreviewNs,
+			state.selectedMatch.col + #state.selectedMatch.matchedText
+		)
+	end
+end
+
+function M.selectPrevMatch()
```

Comment:
selectPrevMatch and selectNextMatch could be merged into one function taking the direction as a parameter

Reply:
are you ok with keeping shorthand functions like this:
``` lua
---@param direction "next" | "prev"
local function cycleMatches(direction)
...
function M.selectNextMatch()
	cycleMatches("next")
end

function M.selectPrevMatch()
	cycleMatches("prev")
end
```

Reply:
why even the wrapper functions? You can just assign `cycleMatches` to the keymap definitions for the popup win?

Reply:
perhaps I'm ignorant here...
As far as I'm aware the only way to do that would be with an inline function since `vim.keymap.set` doesn't have a mechanism to pass extra args to the function.
So If i have to write function()...end, I might as well do it like this instead of the inline function.
That was my thinking process anyway